cpp/src/arrow/util/io-util.cc

```diff
@@ -47,31 +48,26 @@
 #endif
 
 // For filename conversion
-#if defined(_MSC_VER)
+#if defined(_WIN32)
 #include <codecvt>
 #include <locale>
 #include <stdexcept>
 #endif
 
-#if defined(_MSC_VER)
-#define USE_BOOST_FILESYSTEM 1
-#else
-#define USE_BOOST_FILESYSTEM 0
-#endif
-
-#if USE_BOOST_FILESYSTEM
-#include <boost/filesystem.hpp>  // NOLINT
-#endif
+#include <boost/filesystem.hpp>
 
 // ----------------------------------------------------------------------
 // file compatibility stuff
 
-#if defined(__MINGW32__)  // MinGW
-// nothing
-#elif defined(_MSC_VER)  // Visual Studio
+#if defined(_WIN32)
 #include <io.h>
-#else  // POSIX / Linux
-// nothing
+#endif
+
+#ifdef __MINGW32__
+// MinGW lacks this definition
+#ifndef _SH_DENYNO
+#define _SH_DENYNO 0x40
```

Comment:
This is defined in `share.h`.

https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/sopen-s-wsopen-s?view=vs-2019

> The argument shflag is a constant expression that consists of one of the following manifest constants, which are defined in <share.h>.

So we should use `#include <share.h>` instead of `#define _SH_DENYNO 0x40`.

Reply:
Can you check it works?

Reply:
I confirmed this on my local GNU/Linux:

```console
% cat a.c
#include <io.h>

int
main(void)
{
  _SH_DENYNO;
  return 0;
}
% x86_64-w64-mingw32-gcc a.c
a.c: In function 'main':
a.c:6:3: error: '_SH_DENYNO' undeclared (first use in this function)
   _SH_DENYNO;
   ^~~~~~~~~~
a.c:6:3: note: each undeclared identifier is reported only once for each function it appears in
% cat a.c
#include <io.h>
#include <share.h>

int
main(void)
{
  _SH_DENYNO;
  return 0;
}
% x86_64-w64-mingw32-gcc a.c
%